src/ln/channelmanager.rs

```diff
@@ -1465,23 +1465,55 @@ impl ChannelManager {
 	/// still-available channels.
 	fn fail_htlc_backwards_internal(&self, mut channel_state_lock: MutexGuard<ChannelHolder>, source: HTLCSource, payment_hash: &[u8; 32], onion_error: HTLCFailReason) {
 		match source {
-			HTLCSource::OutboundRoute { .. } => {
+			HTLCSource::OutboundRoute { ref route, .. } => {
 				mem::drop(channel_state_lock);
-				if let &HTLCFailReason::ErrorPacket { ref err } = &onion_error {
-					let (channel_update, payment_retryable) = self.process_onion_failure(&source, err.data.clone());
-					if let Some(update) = channel_update {
+				match &onion_error {
+					&HTLCFailReason::ErrorPacket { ref err } => {
+#[cfg(test)]
+						let (channel_update, payment_retryable, onion_error_code) = self.process_onion_failure(&source, err.data.clone());
+#[cfg(not(test))]
+						let (channel_update, payment_retryable, _) = self.process_onion_failure(&source, err.data.clone());
+						if let Some(update) = channel_update {
+							self.channel_state.lock().unwrap().pending_msg_events.push(
+								events::MessageSendEvent::PaymentFailureNetworkUpdate {
+									update,
+								}
+							);
+						}
+						self.pending_events.lock().unwrap().push(
+							events::Event::PaymentFailed {
+								payment_hash: payment_hash.clone(),
+								rejected_by_dest: !payment_retryable,
+#[cfg(test)]
+								error_code: onion_error_code
+							}
+						);
+					},
+					&HTLCFailReason::Reason {
+#[cfg(test)]
+						ref failure_code,
+#[cfg(not(test))]
+						failure_code:_,
+						data:_ } => {
+						// we get a fail_malformed_htlc from the first hop
```

Comment:
"first hop" can we be sure it's the first one and not a previous hop there ?

Reply:
if my understanding is correct, fail_malformed_htlc is converted to fail_htlc by intermediate nodes. so if the origin node received fail_malformed it must be from the first hop in the route.